llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp

```diff
@@ -658,6 +711,9 @@ void DWARFUnit::clearDIEs(bool KeepCUDie) {
   // It depends on the implementation whether the request is fulfilled.
   // Create a new vector with a small capacity and assign it to the DieArray to
   // have previous contents freed.
+  llvm::sys::ScopedWriter FreeLock(CUDieFreeMutex);
+  llvm::sys::ScopedWriter CULock(CUDieArrayMutex);
+  llvm::sys::ScopedWriter AllLock(AllDieArrayMutex);
```

Comment:
I believe taking all 3 locks is overkill, as getting the CUDieFreeMutex writer lock will block until the CUDie and AllDie locks are free. Thoughts?

Reply:
Yeah, sounds about right.

And perhaps once the other two are removed the code'll be basically the same as it was without this patch (except for two lock acquisitions) before and so a lot of the concerns/questions/issues (the newly subdivided functions and names are quite confusing to me, but I've deleted my comments on that so as not to confuse things further if it's all going away anyway) go away?

Reply:
I can't remove the other two locks. They have to stay, as the freeing mechanism requires it. The third lock (Free) is specifically to deal with the issue that @ayermolo pointed out a couple revisions ago. (Freeing could free just the CU DIE while the non-CU DIE stuff was being initialized, resulting in an invalid state of no CU DIE, but non-CU DIE's are parsed). The entire process is kind of a mess, but I need a "CU DIE initialization" RWLock, plus the "Non-CU DIE initialization" lock as two separate entities, as initializing the non-CU DIE's can trigger recursion (you have to be able to grab the reader lock of the CU DIE init to tell that it's already been initialized, then bail)

Reply:
@dwblaikie what are you thoughts here? I've updated the diff to use the 3 locks as efficiently as I can come up with. If you'd like to chat in real time, I'm happy to hop on a VC in whatever host system you're comfortable with (and my calendar is fully blocked off all day Friday) I really do need this diff landed.

Reply:
Sorry for the delays & thanks for the offer to meet more directly (& setting aside time - I didn't see this message until later, that week was a bit busy for me I think... ) - looking into this now.